### PR TITLE
Upgrade/leaflet v1.x

### PR DESCRIPTION
**Pull request by bhaskarvk**
@jcheng5 This PR supersedes #453. On top of #453 I've merged @timelyportfolio's #458. Then merged some other pending PRs #462, #405, #396, #393 which were pending for sometime.

I've verified almost everything from my side. But would appreciate one final round of testing from you and @timelyportfolio. 

After this I'll work to migrate leaflet.extras and leaflet.esri to 1.x, and then we can release all 3 + mapview, mapedit to CRAN simultaneously. I am targeting a CRAN release early 2018, hopefully before rstudio::conf :)

### CODE REVIEW

**Comment by jcheng5** [comment]
Thank you @bhaskarvk and @timelyportfolio for all the hard work on this. I'm in the process of reviewing and will make this a priority this week.

File: R/layers.R [hunk 1]
```diff
@@ -209,6 +211,10 @@ epsg3857 <- "+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y
 #'   the caller's responsibility to ensure that \code{x} is already projected,
 #'   and that \code{extent(x)} is expressed in WGS84 latitude/longitude
 #'   coordinates
+#' @param method the method used for computing values of the new, projected raster image.
+#'   \code{"bilinear"} (the default) is appropriate for continuous data,
+#'   \code{"ngb"} - nearest neighbor - is appropriate for categorical data.
+#'   Ignored if \code{project = FALSE}. See \code{\link{projectRaster}} for details.
```
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply R/layers.R]: I'll give this some thought.

File: R/methods.R [hunk 2]
```diff
@@ -33,11 +33,31 @@ setView <- function(map, lng, lat, zoom, options = list()) {
   )
 }
 
+#' @describeIn map-methods Flys to a given location/zoom-level using smooth pan-zoom.
```
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply R/methods.R]: Good catch. btw flyTo and flyToBounds methods only work on an initialized map (i.e. a map with a center and zoom already set). So technically these methods are shiny only. I'll do some clean up here.

File: R/methods.R [hunk 3]
```diff
@@ -33,11 +33,31 @@ setView <- function(map, lng, lat, zoom, options = list()) {
...
+#' @export
+flyTo <- function(map, lng, lat, zoom, options = list()) {
+  view = evalFormula(list(c(lat, lng), zoom, options))
+
+  dispatch(map,
+    "flyTo",
+    leaflet = {
+      map$x$flyTo = view
+      map$x$fitBounds = NULL
```
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply R/methods.R]: Yeah I think this part needs to be cleaned up a bit.

File: R/methods.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcheng5) [reply R/methods.R]: Oh, OK--I was assuming it'd fly in from 0x0x0. In that case, I'd just throw an error for the local case.

File: inst/errors/errors.Rmd [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# Errors
```
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply inst/errors/errors.Rmd]: Mostly for @timelyportfolio to test and report. Kent, please also remove example files which use leaflet.extras from under inst/examples. If you want I can accept them as PRs on leaflet.extras. It doesn't make sense to have example code with downstream dependencies, unless @jcheng5 disagrees.

File: inst/examples/crosstalk_shiny.R [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    key = ~id,
+    group = "grp1"
+  )
+
+
+  lf <- leaflet(pts) %>%
+    addTiles() %>%
+    addMarkers()
+
+  not_rendered <- TRUE
```
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply inst/examples/crosstalk_shiny.R]: @timelyportfolio ?
https://github.com/rstudio/leaflet/commits/v1.x/inst/examples/crosstalk_shiny.R

File: javascript/src/layer-manager.js [hunk 6]
```diff
@@ -349,7 +349,7 @@ export default class LayerManager {
     if (layerInfo.ctGroup) {
       let ctGroup = this._byCrosstalkGroup[layerInfo.ctGroup];
       let layersForKey = ctGroup[layerInfo.ctKey];
-      let idx = layersForKey ? layersForKey.indexOf(stamp) : -1;
+      let idx = layersForKey ? layersForKey.indexOf(+stamp) : -1;
```
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply javascript/src/layer-manager.js]: This was @cpsievert's fix for #395, part of the PR #396 which I've merged with this PR.

[the remark this answers is not in the document]
Comment (jcheng5) [reply javascript/src/layer-manager.js]: In that case I'd like to undo that conversion to numeric, and instead, convert the stamp to a string at the time that it's created. layer-manager.js#42: `let stamp = L.Util.stamp(layer);` becomes `let stamp = L.Util.stamp(layer) + "";`

File: inst/errors/errors.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (timelyportfolio) [reply inst/errors/errors.Rmd]: Ok, I think I deleted all the tests, experiments, errors that I inadvertently left from my initial work.  Sorry I forgot to remove earlier.

File: inst/examples/crosstalk_shiny.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (timelyportfolio) [reply inst/examples/crosstalk_shiny.R]: should be deleted in ⬆️

File: R/layers.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (tim-salabim) [reply R/layers.R]: To clarify: I basically copied the parameter description from `?projectRaster` from the raster package. I don't think leaflet should try to be smart and decide upon an interpolation method based on input type. In my opinion leaflet should provide a API that is flexible and lets the user decide which method to use. As I alluded to in my comment in #219 there are many instances where integer valued rasters are intended to be categorical. However, I don't think that all users are aware of the fact that there is a `as.factor` method in package raster. This basically boils down to users understanding their data types. 
Regarding color matching, I think something like the following would be sufficient (I simply copied the default settings for `colorFactor` from `?colorFactor`)

```r
method <- match.arg(method, c("bilinear", "ngb"))
if (project) {
  projected <- projectRasterForLeaflet(x, method)
} else {
  projected <- x
}
bounds <- raster::extent(raster::projectExtent(raster::projectExtent(x, crs = sp::CRS(epsg3857)), crs = sp::CRS(epsg4326)))

if (!is.function(colors)) {
  if (method == "ngb") {
    colors <- colorFactor(colors, domain = NULL, levels = NULL, ordered = FALSE,
                          na.color = "#808080", alpha = FALSE, reverse = FALSE)
  } else {
    colors <- colorNumeric(colors, domain = NULL, na.color = "#00000000", alpha = TRUE)
  }
}
```
do the arg matching regardless of the state of argument `project` and use that to decide whether to map colors using `colorNumeric` or `colorFactor`. 

@jcheng5 I am not sure I understand what you mean with your first comment? Where is that scaling being done?

[the remark this answers is not in the document]
Comment (jcheng5) [reply R/layers.R]: > I don't think leaflet should try to be smart and decide upon an interpolation method based on input type. In my opinion leaflet should provide a API that is flexible and lets the user decide which method to use.

What if it's `method = c("auto", "bilinear", "ngb")`? I basically would like to ensure that if it's a factor, it gets ngb by default, and if it's numeric, it gets bilinear by default. But the user will need to tell us with integer if they want ngb.

> @jcheng5 I am not sure I understand what you mean with your first comment? Where is that scaling being done?

It's done on the fly in JS as the canvas is rendered: https://github.com/bhaskarvk/leaflet/blob/fb7bc8178baab3faf662a07acfe060555b5d8bb4/javascript/src/methods.js#L1139-L1237

File: javascript/src/layer-manager.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (bhaskarvk) [reply javascript/src/layer-manager.js]: I wouldn't do that. L.Util.stamp is used in some other places too (including leaflet.exttras) and I don't want to introduce additional complexity by changing it's type from the default numeric to string. I think I know what the problem is and how to fix it.

File: R/layers.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (tim-salabim) [reply R/layers.R]: `method = c("auto", "bilinear", "ngb")` is a good idea! 
Thanks for the clarification on the scaling. Just to be completely sure I understand correctly, with server-side you mean before passing to JS? Hence, client is the JS side?

[the remark this answers is not in the document]
Comment (jcheng5) [reply R/layers.R]: > Just to be completely sure I understand correctly, with server-side you mean before passing to JS? Hence, client is the JS side?

Yes and yes. Sorry for any confusion.

**Pull request closed without merging**